src/js/view/controls/components/menu/menu.js

```diff
@@ -0,0 +1,499 @@
+        this.openMenus = [];
+        this.items = [];
+        this.visible = false;
+        this.parentMenu = _parentMenu;
+        this.mainMenu = !this.parentMenu ? this : this.parentMenu.mainMenu;
+        this.closeButton = (this.parentMenu && this.parentMenu.closeButton) || this.createCloseButton(_localization);
+        this.categoryButton = null;
+        if (this.isSubmenu) {
+            if (this.parentMenu.name === 'settings') {
+                this.createCategoryButton(_localization);
+            }
+            this.parentMenu.appendMenu(this);
+        } else {
+            this.ui = addGlobalMenuKeyListener(this);
+        }
+        this.onDocumentClick = this.onDocumentClick.bind(this);
+        this.open = this.open.bind(this);
+        this.close = this.close.bind(this);
+        this.toggle = this.toggle.bind(this);
+    }
+    createCloseButton(localization) {
+        const closeButton = button('jw-settings-close', (e) => this.close(e), localization.close, [cloneIcon('close')]);
+        this.topbar.appendChild(closeButton.element());
+        closeButton.show();
+        closeButton.ui.on('keydown', (e) => {
+            const sourceEvent = e.sourceEvent;
+            const key = sourceEvent.key.replace(/(Arrow|ape)/, '');
+            // Close settings menu when enter is pressed on the close button
+            // or when tab or right arrow key is pressed since it is the last element in topbar
+            if (key === 'Enter' || key === 'Right' || (key === 'Tab' && !sourceEvent.shiftKey)) {
+                this.close(e);
+            }
+        }, this);
+        this.topbar.appendChild(closeButton.element());
+        return closeButton;
+    }
+    createCategoryButton(localization) {
+        const localizationKeys = {
+            captions: 'cc',
+            audioTracks: 'audioTracks',
+            quality: 'hd',
+            playbackRates: 'playbackRates',
+        };
+        let localizedName = localization[localizationKeys[this.name]];
+        if (this.name === 'sharing') {
+            localizedName = localization.sharing.heading;
+        }
+        this.categoryButton = categoryButton(this, localizedName);
+        this.categoryButton.element().setAttribute('name', this.name);
+    }
+    createBackButton(localization) {
+        const backButton = button('jw-settings-close', this.close, localization.close, [cloneIcon('close')]);
+        return backButton;
+    }
+    createTopbar() {
+        const topbar = createElement(`<div class="jw-submenu-topbar"></div>`);
+        this.topbar = topbar;
+        prependChild(topbar, this.el);
+    }
+    createItems(genericItems, action, options = {}, Item = RadioMenuItem) {
+        const itemType = this.name;
+        const categoryButtonElement = this.categoryButton.element();
+        const menuItems = genericItems.map((item, index) => {
+            let content;
+            let argument;
+            switch (itemType) {
+                case 'quality':
```

Comment:
what are the benefits of this approach vs a subclass pattern where each menu type implements it's own items. i think that logical breakdown makes more sense but this is of course more concise